From 447c77d29522dd2f3978647b6c5713c256c399cf Mon Sep 17 00:00:00 2001
From: Benjamin Coddington <bcodding@redhat.com>
Date: Wed, 8 Oct 2014 19:46:08 -0400
Subject: [fs] lockd: Try to reconnect if statd has moved

Message-id: <12359478168126775b7559239d5f7b0dd4afa0a8.1412792619.git.bcodding@redhat.com>
Patchwork-id: 97279
O-Subject: [RHEL7.1 PATCH fs 2/2] lockd: Try to reconnect if statd has moved
Bugzilla: 1120850
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

If rpc.statd is restarted, upcalls to monitor hosts can fail with
ECONNREFUSED.  In that case force a lookup of statd's new port and retry the
upcall.

Signed-off-by: Benjamin Coddington <bcodding@redhat.com>
Cc: stable@vger.kernel.org
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 173b3afceebe76fa2205b2c8808682d5b541fe3c
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1120850
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/lockd/mon.c b/fs/lockd/mon.c
index 1812f02..6ae664b 100644
--- a/fs/lockd/mon.c
+++ b/fs/lockd/mon.c
@@ -159,6 +159,12 @@ static int nsm_mon_unmon(struct nsm_handle *nsm, u32 proc, struct nsm_res *res,
 
  msg.rpc_proc = &clnt->cl_procinfo[proc];
  status = rpc_call_sync(clnt, &msg, RPC_TASK_SOFTCONN);
+ if (status == -ECONNREFUSED) {
+  dprintk("lockd: NSM upcall RPC failed, status=%d, forcing rebind\n",
+    status);
+  rpc_force_rebind(clnt);
+  status = rpc_call_sync(clnt, &msg, RPC_TASK_SOFTCONN);
+ }
  if (status < 0)
   dprintk("lockd: NSM upcall RPC failed, status=%d\n",
     status);
-- 
1.7.1